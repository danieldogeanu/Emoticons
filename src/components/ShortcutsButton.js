import React, { Component } from 'react';
import { animateCSS } from '../animate';
import Icon from './Icon';
import '../styles/components/ShortcutsButton.scss';

class ShortcutsButton extends Component {
	componentDidMount() {
		this.ShortcutsScreen = document.querySelector('.ShortcutsScreen');
		this.ShortcutsButton = document.querySelector('.ShortcutsButton');
	}

	handleClick() {
		animateCSS(this.ShortcutsScreen, 'fadeIn');
		this.ShortcutsScreen.classList.add('show');
		this.ShortcutsButton.classList.add('hide');
	}

	render() {
		const shortcutsButtonText = 'Keyboard Shortcuts';
		return (
			<button className="ShortcutsButton" 
				title={shortcutsButtonText}
				onClick={() => this.handleClick()}>
				<span className="show-for-screen-reader">
					{shortcutsButtonText}
				</span>
				<Icon name="keyboard" />
			</button>
		);
	}
}

export default ShortcutsButton;