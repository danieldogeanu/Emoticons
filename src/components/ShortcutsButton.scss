@import '../styles/includes';

.ShortcutsButton {
	display: block;
	position: fixed;
	left: rem(40);
	bottom: rem(30);
	width: auto;
	height: auto;
	border: 0 none;
	border-radius: rem(4);
	background-color: transparent;
	transition: color 0.3s ease, opacity 0.5s ease;
	color: color-scheme(scrollup, normal);
	cursor: pointer;
	&:hover { color: color-scheme(scrollup, hover); }
	&:focus, &:active { outline: none; }
	&.hide { display: none; }
	.icon.keyboard {
		width: rem(45);
		height: rem(45);
	}
}