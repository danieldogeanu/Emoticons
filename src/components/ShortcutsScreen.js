import React, { Component } from 'react';
import Icon from './Icon';
import './ShortcutsScreen.scss';

class ShortcutsClose extends Component {	
	handleClick() {
		const shortcutsScreen = document.querySelector('.ShortcutsScreen');
		const shortcutsButton = document.querySelector('.ShortcutsButton');
		shortcutsScreen.classList.remove('show');
		shortcutsButton.classList.remove('hide');
	}

	render() {
		const shortcutsCloseText = 'Close Shortcuts Screen';
		return (
			<button className="ShortcutsClose" 
				title={shortcutsCloseText}
				onClick={() => this.handleClick()}>
				<span className="show-for-screen-reader">
					{shortcutsCloseText}
				</span>
				<Icon name="close" />
			</button>
		);
	}
}

class ShortcutsScreen extends Component {
	render() {
		const shortcuts = [
			{ key: 's', description: 'Selects Search Input' },
			{ key: 'esc', description: 'Clears Search Input / Exits Menu' },
			{ key: 'k', description: 'Opens Keyboard Shortcuts Screen' },
			{ key: 't', description: 'Scrolls Back Up' },
		];
		return (
			<div className="ShortcutsScreen">
				<div className="shortcutsList">
					<span className="shortcutsTitle">Keyboard Shortcuts</span>
					<ul>
						{shortcuts.map(shortcut => {
							return <li key={shortcut.key}><kbd>{shortcut.key}</kbd> {shortcut.description}</li>;
						})}
					</ul>
					<ShortcutsClose />
				</div>
			</div>
		);
	}
}

export default ShortcutsScreen;