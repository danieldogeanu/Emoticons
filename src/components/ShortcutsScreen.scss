@import '../styles/includes';

// TODO: Add Animation for .ShortcutsScreen on Opening.

.ShortcutsScreen {
	display: none;
	position: fixed;
	top: 0; right: 0;
	bottom: 0; left: 0;
	padding: rem(30);
	@include linear-gradient(left, 
		color-scheme(shortcuts, gradient-start), rem(200), 
		color-scheme(shortcuts, gradient-end), 100%
	);
	@include breakpoint(min, rem(481)) { padding: rem(40); }
	&.show { display: block; }

	.shortcutsList {
		position: absolute;
		left: rem(30);
		bottom: rem(100);
		@include breakpoint(min, rem(481)) { left: rem(40); }
	}

	.shortcutsTitle {
		display: block;
		width: 100%;
		margin: 0 0 rem(20) 0;
		font-weight: weight(extra-bold);
		font-size: rem(14);
		text-transform: uppercase;
	}

	ul {
		display: block;
		list-style: none;
		margin: 0; padding: 0;
		li {
			margin: rem(12) 0;
			font-size: rem(16);
			&:hover kbd {
				border: rem(2) solid color-scheme(shortcuts, hover);
				color: color-scheme(shortcuts, hover);
			}
		}
	}

	kbd {
		display: inline-block;
		width: rem(45);
		margin: 0 rem(10) 0 0;
		padding: rem(6) rem(10);
		border-radius: rem(4);
		border: rem(2) solid color-scheme(shortcuts, normal);
		background-color: transparent;
		color: color-scheme(shortcuts, normal);
		font-family: font(open-sans);
		font-weight: weight(extra-bold);
		font-size: rem(12);
		line-height: 1;
		text-transform: uppercase;
		transition: all 0.3s ease;
		&:hover {
			border: rem(2) solid color-scheme(shortcuts, hover);
			color: color-scheme(shortcuts, hover);
		}
	}
}

.ShortcutsClose {
	display: flex;
	justify-content: flex-start;
	align-items: center;
	position: fixed;
	left: rem(30);
	bottom: rem(31);
	width: rem(45);
	height: rem(30);
	border: 0 none;
	border-radius: rem(4);
	background-color: transparent;
	transition: color 0.3s ease, opacity 0.5s ease;
	color: color-scheme(scrollup, normal);
	cursor: pointer;
	@include breakpoint(min, rem(481)) { left: rem(40); bottom: rem(41); }
	&:hover { color: color-scheme(scrollup, hover); }
	&:focus, &:active { outline: none; }
	.icon.close {
		width: rem(26);
		height: rem(26);
	}
}