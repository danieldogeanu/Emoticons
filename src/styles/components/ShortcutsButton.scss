@import '../includes';

.ShortcutsButton {
	cursor: pointer;
	display: none;
	position: fixed;
	left: rem(30);
	bottom: rem(20);
	width: auto;
	height: auto;
	border: 0 none;
	border-radius: rem(4);
	background-color: transparent;
	color: color-scheme(scrollup, normal);
	transition: color 0.3s ease, opacity 0.5s ease;
	@include breakpoint(min, rem(481)) { left: rem(40); bottom: rem(30); }
	@include breakpoint(min, rem(961)) { display: block; }
	&:hover { color: color-scheme(scrollup, hover); }
	&:focus, &:active { outline: none; }
	&.hide { display: none; }
	.icon.keyboard {
		width: rem(45);
		height: rem(45);
	}
}

// TODO: Maybe include ShortcutsButton on a black bar at the bottom of the screen on mobile along with ScrollUp button.