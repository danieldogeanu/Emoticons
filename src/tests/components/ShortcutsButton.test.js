import React from 'react';
import {render, fireEvent, cleanup} from '@testing-library/react';
import ShortcutsButton from '../../components/ShortcutsButton';
import ShortcutsScreen from '../../components/ShortcutsScreen';

jest.mock('../../components/ShortcutsScreen', () => () => (
	<div className="ShortcutsScreen" data-testid="ShortcutsScreen"></div>
));

const btnTitle = 'Keyboard Shortcuts';

describe('ShortcutsButton Component', () => {

	afterEach(cleanup);

	it('renders shortcuts button properly', () => {
		const {getByTitle, getByTestId, container} = render(<ShortcutsButton />);
		const renderedBtn = getByTitle(btnTitle);
		const renderedSRText = getByTestId('ScreenReaderText');
		const renderedIcon = container.querySelector('svg');

		expect(renderedBtn).toBeInTheDocument();
		expect(renderedBtn).toHaveClass('ShortcutsButton');
		expect(renderedBtn).toHaveAttribute('title', btnTitle);
		expect(renderedBtn).toContainElement(renderedSRText);
		expect(renderedBtn).toContainElement(renderedIcon);
		expect(renderedBtn).toMatchSnapshot();

		expect(renderedSRText).toHaveTextContent(btnTitle);
		expect(renderedIcon).toHaveClass('icon', 'keyboard');
	});

	it('opens shortcuts screen', () => {
		const {getByTitle, getByTestId} = render(
			<div data-testid="test-parent">
				<ShortcutsButton />
				<ShortcutsScreen />
			</div>
		);
		const renderedBtn = getByTitle(btnTitle);
		const renderedScreen = getByTestId('ShortcutsScreen');

		fireEvent.click(renderedBtn);

		expect(renderedBtn).toHaveClass('ShortcutsButton', 'hide');
		expect(renderedScreen).toHaveClass('ShortcutsScreen', 'animated', 'fadeIn', 'show');
	});

});
