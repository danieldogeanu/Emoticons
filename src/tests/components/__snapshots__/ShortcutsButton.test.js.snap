// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShortcutsButton Component renders shortcuts button properly 1`] = `
<button
  class="ShortcutsButton"
  title="Keyboard Shortcuts"
>
  <span
    class="show-for-screen-reader"
    data-testid="ScreenReaderText"
  >
    Keyboard Shortcuts
  </span>
  <svg
    class="icon keyboard"
  >
    <use
      xlink:href="#keyboard"
    />
  </svg>
</button>
`;
